lib/streamlit/errors.py

```diff
@@ -133,3 +137,227 @@ def __init__(self, module_name, *args):
+    def __init__(self, message: str, **kwargs):
+        exception_name = type(self).__name__
+        exception_message = gettext.gettext(exception_name)
+        if exception_message == exception_name:
+            # gettext returns the message ID if there is no translation.
+            # In this case, we want to use the message provided in the constructor.
+            exception_message = message
+        super().__init__((exception_message).format(**kwargs))
+        self._exec_kwargs = kwargs
+
+    @property
+    def exec_kwargs(self) -> dict[str, Any]:
+        return self._exec_kwargs
+
+
+# st.number_input
+class StreamlitNumberInputDifferentTypesError(LocalizableStreamlitException):
+    """Exception raised mixing floats and ints in st.number_input."""
+
+    def __init__(
+        self,
+        value: int | float,
+        min_value: int | float,
+        max_value: int | float,
+        step: int | float,
+    ):
+        error_message = "All numerical arguments must be of the same type."
+
+        if value:
+            value_type = type(value).__name__
+            error_message += f"\n`value` has {value_type} type."
+
+        if min_value:
+            min_value_type = type(min_value).__name__
+            error_message += f"\n`min_value` has {min_value_type} type."
+
+        if max_value:
+            max_value_type = type(max_value).__name__
+            error_message += f"\n`max_value` has {max_value_type} type."
+
+        if step:
+            step_type = type(step).__name__
+            error_message += f"\n`step` has {step_type} type."
+
+        super().__init__(error_message)
+
+
+class StreamlitInputInvalidValueError(LocalizableStreamlitException):
+    """Exception raised when the `min_value` is greater or the `max_value` is less than the `value` in `st.number_input`."""
+
+    def __init__(
+        self, value: int | float, min_value: int | float, max_value: int | float
+    ):
+        error_message = None
+
+        if min_value is not None and value < min_value:
+            error_message = (
+                f"The `value` {value} is less than the `min_value` {min_value}."
+            )
+        elif max_value is not None and value > max_value:
+            error_message = (
+                f"The `value` {value} is greater than the `max_value` {max_value}."
+            )
+
+        super().__init__(error_message)
+
+
+class StreamlitJSNumberBoundsError(LocalizableStreamlitException):
+    """Exception raised when a number exceeds the Javascript limits."""
+
+    def __init__(self, message: str):
+        super().__init__(message)
+
+
+class StreamlitNumberInputInvalidFormatError(LocalizableStreamlitException):
+    """Exception raised when the format string for `st.number_input` contains
+    invalid characters.
+    """
+
+    def __init__(self, format: str):
+        super().__init__(
+            f"Format string for st.number_input contains invalid characters: {format}"
```

Comment:
I think it's essential to separating message template and template arguments at this level to eventually allow localizing or adapting messages on different platforms.

In this case:
 
```python
super().__init__(
     "Format string for st.number_input contains invalid characters: {format}",
      format=format
 )
```

If you would want to localize the message, you can just add a template that uses the same variable placeholders. For example, translation to German:

```
"Format string für st.number_input beinhaltet ungültige Zeichen: {format}"
```

This also allows us to provide options to localize or adapt messages via something like gettext and/or just by allowing users custom handling, e.g.:

```python
def exception_handler(exc: BaseException) -> BaseException | None:
    if isinstance(exc, StreamlitNumberInputInvalidFormatError):
        # Custom error handling
        st.error(
           "Format string für st.number_input beinhaltet ungültige Zeichen: {format}".format(**exc.exec_kwargs),
        )
        return None
    return exc
 ```

Reply:
Ah yes, good point. I'll fix that.